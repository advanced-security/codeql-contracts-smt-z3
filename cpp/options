semmle-extractor-options:--clang -std=c++14 -nostdinc++ -I../../../../include