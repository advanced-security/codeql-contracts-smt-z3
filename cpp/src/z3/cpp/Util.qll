import cpp
import z3.Z3

predicate supportedPrimitiveType(Type t, Z3::Sort s) {
  t instanceof BoolType and s instanceof Z3::BoolSort
  or
  t instanceof IntType and s instanceof Z3::IntSort
}

Z3::Sort z3SortForQlType(Type t, VariableAccess va) {
  va.getTarget().getType() = t and
  (
    t instanceof PointerType and dereferenced(va) and supportedPrimitiveType(t.stripType(), result)
    or
    (t instanceof PointerType and not dereferenced(va)) and
    result instanceof Z3::OptionNullSort
    or
    supportedPrimitiveType(t, result)
  )
}

Z3::Sort z3SortForQlType(VariableAccess va) {
  result = z3SortForQlType(va.getTarget().getType(), va)
}

string z3TypeForQlType(Type t, VariableAccess va) { z3SortForQlType(t, va).toString() = result }

string z3TypeForQlRawType(Type t) { z3SortForQlRawType(t).toString() = result }

Z3::Sort z3SortForQlRawType(Type t) {
  t instanceof PointerType and result instanceof Z3::OptionNullSort
  or
  supportedPrimitiveType(t, result)
}
