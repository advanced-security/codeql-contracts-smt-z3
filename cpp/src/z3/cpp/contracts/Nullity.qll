import cpp
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis
import semmle.code.cpp.controlflow.Nullness
import semmle.code.cpp.controlflow.Guards

string describe(VariableAccess va) {
  // Note: fields are not analyzed, as nullness checks don't work well on them.
  lowerBound(va) > 0 and result = "range analysis non-null"
  or
  upperBound(va) <= 0 and result = "range analysis null"
  or
  exists(GuardCondition guard, Expr left, Expr right, int k, BasicBlock block |
    guard.ensuresEq(left, right, k, block, true) and
    va.getEnclosingBlock() = block and
    left.(VariableAccess).getTarget() = va.getTarget() and
    right.getValue() = "0" and
    result = "guards == " + k.toString()
    or
    guard.ensuresEq(left, right, k, block, false) and
    va.getEnclosingBlock() = block and
    left.(VariableAccess).getTarget() = va.getTarget() and
    right.getValue() = "0" and
    result = "guards != " + k.toString()
  )
}

predicate possiblyNonNull(VariableAccess va) {
  lowerBound(va) > 0
  or
  exists(GuardCondition guard, Expr left, Expr right, int k, BasicBlock block |
    guard.ensuresEq(left, right, k, block, false) and
    va.getEnclosingBlock() = block and
    left.(VariableAccess).getTarget() = va.getTarget() and
    right.getValue() = "0"
  )
}

predicate possiblyNull(VariableAccess va) { not possiblyNonNull(va) }
