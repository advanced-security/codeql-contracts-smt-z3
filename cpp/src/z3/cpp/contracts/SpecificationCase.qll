/**
 * A Specification consists of specification cases which in turn consist of
 * specification clauses.
 */

import cpp
import semmle.code.cpp.commons.Assertions
import z3.cpp.Translation

/**
 * A `SpecifiedElement` is an actual program level element that either
 * concretely or synthetically is attached to a specification element.
 *
 * In the case of embedded specifications, they are always attached to the
 * `Function` in which they reside. In the case of others, e.g., macro level
 * assertions, they are attached to the macro. It is important to be careful
 * about this point. We define the rest of our elements so that a single
 * program element can have one or more specification elements attached; this is
 * necessary to be consistent between embedded and non-embedded specifications.
 */
abstract class SpecifiedElement extends Element {
  abstract Expr getBody();

  abstract Function getFunction();

  abstract string getKeyword();

  abstract string getSource();

  abstract predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  );

  // it's worth noting that this is modeled as string mainly because
  // of embedded specifications. The embedded specifications make it hard
  // to store the abstract representations. See `SMT.qll` for some abstract
  // classes for working with SMT expressions.
  string getSMTDeclarations() { result = declarationsForExpr(getBody()) }

  string getSMTAsserts() { result = exprToZ3(getBody()) }
}

/** A SpecificationClause may be one of a set of types of SpecificationClauses. */
private newtype TCodeQLSpecificationClauseType =
  TCodeQLSpecificationClauseRequires(SpecifiedElement e) { e.getKeyword() = "requires" } or
  TCodeQLSpecificationClauseEnsures(SpecifiedElement e) { e.getKeyword() = "ensures" } or
  TCodeQLSpecificationClauseSignals(SpecifiedElement e) { e.getKeyword() = "signals" } or
  TCodeQLSpecificationClauseSignalsOnly(SpecifiedElement e) { e.getKeyword() = "signals_only" } or
  TCodeQLSpecificationClauseWhen(SpecifiedElement e) { e.getKeyword() = "when" } or
  TCodeQLSpecificationClauseAssignable(SpecifiedElement e) { e.getKeyword() = "assignable" } or
  TCodeQLSpecificationClauseCallable(SpecifiedElement e) { e.getKeyword() = "callable" }

/**
 * A SpecificationCase may consist of either a set of SpecificationClauses or a
 * SpecificationCase.
 */
private newtype TCodeQLSpecificationCaseType =
  TCodeQLSingleSpecificationCase(Function f) { exists(SpecifiedElement e | e.getFunction() = f) } or
  TCodeQLEmptySpecificationCase()

abstract class CodeQLSpecificationCase extends TCodeQLSpecificationCaseType {
  Function f;

  abstract CodeQLSpecificationClause getASpecificationClause();

  string toString() { result = "specification of " + f.getName() }

  Function getFunction() { result = f }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    f.getLocation().hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }
}

class CodeQLSingleSpecificationCase extends CodeQLSpecificationCase, TCodeQLSingleSpecificationCase {
  CodeQLSingleSpecificationCase() { this = TCodeQLSingleSpecificationCase(f) }

  override CodeQLSpecificationClause getASpecificationClause() {
    exists(CodeQLSpecificationClause c |
      c.getFunction() = f and
      result = c
    )
  }
}

abstract class CodeQLSpecificationClause extends TCodeQLSpecificationClauseType {
  Function function;
  SpecifiedElement e;

  string toString() { result = getKeyword() + " " + e.getSource() }

  Function getFunction() { result = function }

  Expr getClauseExpr() { result = e.getBody() }

  string getSMTDeclarations() { result = e.getSMTDeclarations() }

  string getSMTAsserts() { result = e.getSMTAsserts() }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    e.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  string getKeyword() { result = e.getKeyword() }
}

class CodeQLSpecificationClauseRequires extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseRequires
{
  CodeQLSpecificationClauseRequires() {
    this = TCodeQLSpecificationClauseRequires(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseEnsures extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseEnsures
{
  CodeQLSpecificationClauseEnsures() {
    this = TCodeQLSpecificationClauseEnsures(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseSignals extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseSignals
{
  CodeQLSpecificationClauseSignals() {
    this = TCodeQLSpecificationClauseSignals(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseSignalsOnly extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseSignalsOnly
{
  CodeQLSpecificationClauseSignalsOnly() {
    this = TCodeQLSpecificationClauseSignalsOnly(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseWhen extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseWhen
{
  CodeQLSpecificationClauseWhen() {
    this = TCodeQLSpecificationClauseWhen(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseAssignable extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseAssignable
{
  CodeQLSpecificationClauseAssignable() {
    this = TCodeQLSpecificationClauseAssignable(e) and
    function = e.getFunction()
  }
}

class CodeQLSpecificationClauseCallable extends CodeQLSpecificationClause,
  TCodeQLSpecificationClauseCallable
{
  CodeQLSpecificationClauseCallable() {
    this = TCodeQLSpecificationClauseCallable(e) and
    function = e.getFunction()
  }
}

private newtype TCodeQLSpecificationStmtType =
  TCodeQLSpecificationStmtAssume(SpecifiedElement e) { e.getKeyword() = "assume" }

abstract class CodeQLSpecificationStmt extends TCodeQLSpecificationStmtType {
  Function function;
  SpecifiedElement e;

  string toString() { result = getKeyword() + " " + e.getSource() }

  Function getFunction() { result = function }

  Expr getExpr() { result = e.getBody() }

  string getSMTDeclarations() { result = e.getSMTDeclarations() }

  string getSMTAsserts() { result = e.getSMTAsserts() }

  predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    e.hasLocationInfo(filepath, startline, startcolumn, endline, endcolumn)
  }

  string getKeyword() { result = e.getKeyword() }
}

class CodeQLSpecificationStmtAssume extends CodeQLSpecificationStmt, TCodeQLSpecificationStmtAssume {
  CodeQLSpecificationStmtAssume() {
    this = TCodeQLSpecificationStmtAssume(e) and
    function = e.getFunction()
  }
}
