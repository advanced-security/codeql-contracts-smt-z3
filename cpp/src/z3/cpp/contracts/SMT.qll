import z3.cpp.contracts.SpecificationCase
import semmle.code.cpp.rangeanalysis.SimpleRangeAnalysis

module Contracts {
  bindingset[exp, lb, ub]
  private string boundsToAssert(string exp, float lb, float ub) {
    result = upperBoundsToAssert(exp, ub) + "\n" + lowerBoundsToAssert(exp, lb) + "\n"
  }

  bindingset[exp, ub]
  private string upperBoundsToAssert(string exp, float ub) {
    result = "(assert (<= " + exp + " " + ub + ")) ; establishment of upper bounds"
  }

  bindingset[exp, lb]
  private string lowerBoundsToAssert(string exp, float lb) {
    result = "(assert (>= " + exp + " " + lb + ")) ; establishment of lower bounds"
  }

  bindingset[exp, categoricalValue]
  private string categoricalValueToAssert(string exp, string categoricalValue) {
    result = "(assert (= " + exp + " " + categoricalValue + ")) ; establishment categorical value"
  }

  /**
   * This predicate works differently than the way that preconditions are constructed.
   * In the case of preconditions the scope is limited to the precondition's scope.
   * The constants and declarations however are exposed at the script level. For
   * this reason we must de-duplicate the declarations.
   *
   * There are two types of declarations we include. Those that are directly needed
   * in the scope of the contract and those that are needed in the body of the
   * callsite.
   */
  string extractDeclarations(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // handles the declarations inside of the specified method, including
    // the ones stored in the spec database.
    specCase.getASpecificationClause().getSMTDeclarations() = result
    or
    // handles the declarations possibly needed by injecting scoped
    // variables
    fc.getTarget() = specCase.getFunction() and
    // analyze each argument
    exists(
      BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Z3::Sort z3Sort, Expr valueExpr,
      Expr anchorExpr, string id
    |
      fc = bfc and
      valueExpr = bfc.getAValueExpression(anchorExpr, p) and
      // pull the type from the parameter if it's the value
      // being passed to the function
      (
        if valueExpr = anchorExpr
        then (
          z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
        ) else (
          z3Sort = z3SortForQlType(valueExpr) and
          id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
        )
      ) and
      result = "(declare-const " + id + " " + z3Sort.toString() + ")"
    )
  }

  string concatDeclarations(CodeQLSpecificationCase specCase, FunctionCall fc) {
    fc.getTarget() = specCase.getFunction() and
    result = concat(string s | s = extractDeclarations(specCase, fc) | s + "\n")
  }

  string assertPreconditionBody(CodeQLSpecificationClause clause) {
    result = concat(CodeQLSpecificationClause c | c = clause | (c.getSMTAsserts() + "\n"))
  }

  string getPrecondition(CodeQLSpecificationCase specCase) {
    exists(int preconditionNumber, string pcBody |
      pcBody =
        rank[preconditionNumber](string pc |
          pc = assertPreconditionBody(specCase.getASpecificationClause())
        |
          pc
        ) and
      result =
        "(define-fun precondition_" + preconditionNumber.toString() + "() Bool \n\t" + pcBody +
          ")\n\n"
    )
  }

  string concatPreconditions(CodeQLSpecificationCase specCase) {
    result = concat(getPrecondition(specCase) + "\n")
  }

  /**
   * Boundary analysis is done on a per argument basis to the function call.
   * We do this by mapping the formal parameter names in SMT to the values
   * produced by range analysis at the call site for its actual parameters.
   */
  string getBounds(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // link these together
    fc.getTarget() = specCase.getFunction() and
    // analyze each argument
    exists(
      BoundaryAnalyzedFunctionCall bfc, ContractParameter p, Z3::Sort z3Sort, Expr valueExpr,
      Expr anchorExpr, string id
    |
      fc = bfc and
      valueExpr = bfc.getAValueExpression(anchorExpr, p) and
      // pull the type from the parameter if it's the value
      // being passed to the function
      (
        if valueExpr = anchorExpr
        then (
          z3Sort = z3SortForQlRawType(p.getType()) and id = p.toId()
        ) else (
          z3Sort = z3SortForQlType(valueExpr) and
          id = valueExpr.(TerminalVariableAccess).toId(anchorExpr, p)
        )
      ) and
      // it isn't necessary to break up these cases --
      // we could just take a disjunction but this is presented
      // this way for clarity that we are taking different approaches
      // to value analysis.
      if z3Sort = Z3::TOptionNullSort()
      then
        exists(string categoricalValue |
          bfc.getACategoricalValue(valueExpr, anchorExpr, p) = categoricalValue and
          result = categoricalValueToAssert(id, categoricalValue)
        )
      else
        exists(float lb, float ub |
          bfc.getALowerBound(valueExpr.getParent*(), anchorExpr, p) = lb and
          ub = bfc.getAUpperBound(valueExpr.getParent*(), anchorExpr, p) and
          result = boundsToAssert(id, lb, ub)
        )
    )
  }

  string concatBounds(CodeQLSpecificationCase specCase, FunctionCall fc) {
    fc.getTarget() = specCase.getFunction() and
    result = concat(string s | s = getBounds(specCase, fc) | s + "\n")
  }

  /**
   * The goal of contract checking is not to find an assignment of variables
   * that allows a formula to be true; the goal is to find an assignment wherein
   * the formula is false. Therefore to make this work with a SAT solver
   * (which finds assignments that make a formula true) we use it in reverse
   * and ask the solver for a cases wherein the solver searches for an
   * assignment to an invalid form of the formula.
   *
   * Stated more concisely, for a formula $F$ over variables $V$ a counter
   * example is an assignment of variables $V$ for which $\neg F$ is satisfiable.
   */
  string getCounterExamples(CodeQLSpecificationCase specCase) {
    exists(string preconditionList, int numberOfPreconditions |
      // note -- we translate to the precondition body here because it is
      // possible to have duplicate preconditions. These are automatically
      // filtered out by set equality so to replicate that behavior here,
      // the conversion via `assertPreconditionBody` ensures that the
      // deduplication takes place.
      numberOfPreconditions = count(assertPreconditionBody(specCase.getASpecificationClause())) and
      preconditionList =
        concat(int idx | idx = [1 .. numberOfPreconditions] | "precondition_" + idx.toString() + " ")
            .trim() and
      if numberOfPreconditions = 1
      then result = "(assert (not " + preconditionList + "))"
      else result = "(assert (not (and " + preconditionList + ")))"
    )
  }

  string toSMT(CodeQLSpecificationCase specCase, FunctionCall fc) {
    // scope this down to a single function
    fc.getTarget() = specCase.getFunction() and
    // do the boundary analysis
    result =
      "                     \n" + ";;                            \n" +
        ";; Declarations               \n" +
        "(declare-datatypes () ((OptionNullSort MaybeNull NotNull)))\n\n" +
        concatDeclarations(specCase, fc) + ";;                            \n" +
        ";; Precondition Definition    \n" + concatPreconditions(specCase) +
        ";;                            \n" + ";; Boundary/Context Assertions\n" +
        concatBounds(specCase, fc) + ";;                            \n" +
        ";; Counter Example Assertion  \n" + getCounterExamples(specCase)
  }
}
