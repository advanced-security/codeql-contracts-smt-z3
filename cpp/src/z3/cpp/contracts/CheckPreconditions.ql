/**
 * @id cpp/contracts/check-preconditions
 * @name Check Preconditions
 * @description Check the preconditions of called methods and report possible
 *              violations.
 * @kind problem
 * @precision high
 * @problem.severity warning
 * @tags cpp/contracts/check-preconditions
 *       codeql/contracts/preconditions/automated
 */

import cpp
import z3.cpp.contracts.SpecificationCase
import z3.cpp.contracts.SMT
import z3.Z3
// enable comment specs
import z3.cpp.contracts.style.AnnotatedCAssertStyleSpecifiedElement

from FunctionCall fc, CodeQLSpecificationCase specCase, string proof, string ce
where
  specCase.getFunction() = fc.getTarget() and
  proof = Contracts::toSMT(specCase, fc) and
  ce = Z3::getModel(proof)
select fc,
  "The $@ at guarding this function call may not be satisfied by this implementation. The SMT Solver found the following model which violates this method's precondition: \n\n"
    + ce, specCase, "specification case"
