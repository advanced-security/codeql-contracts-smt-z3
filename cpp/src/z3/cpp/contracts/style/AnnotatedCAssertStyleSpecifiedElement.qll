import cpp
import z3.cpp.contracts.SpecificationCase

/**
 * Describes a C-style assert that has been annotated with a comment indicating
 * it should be treated as a contract specification.
 *
 * For example:
 *
 * ```
 *   /*@ requires @* /
 *   assert(x > 0);
 * ```
 */
class AnnotatedCAssertStyleSpecifiedElement extends SpecifiedElement, MacroInvocation {
  Expr body;
  string keyword;
  string contents;
  Comment exprComment;

  AnnotatedCAssertStyleSpecifiedElement() {
    // model the LHS of a logical or style CAssert
    exists(LogicalOrExpr b |
      body = b.getLeftOperand() and
      b = this.getAGeneratedElement() and
      not exists(LogicalOrExpr e1 |
        not e1 = b and
        b.getParent*() = e1
      )
    ) and
    exprComment.getCommentedElement() = this.getExpr().getEnclosingStmt() and
    (
      // Find comments such as '/*@ requires @*/' or '/*@ ensures @*/'.
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+requires\\s+@\\*\\/$", _, _) and
      keyword = "requires"
      or
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+ensures\\s+@\\*\\/$", _, _) and
      keyword = "ensures"
      or
      contents =
        exprComment
            .getContents()
            .splitAt("\n")
            .regexpFind("^\\s*\\/\\*@\\s+assume\\s+@\\*\\/$", _, _) and
      keyword = "assume"
    )
  }

  override string getSource() { getBody().toString() = result }

  override Expr getBody() { result = body }

  override Function getFunction() { result = getBody().getEnclosingFunction() }

  override predicate hasLocationInfo(
    string filepath, int startline, int startcolumn, int endline, int endcolumn
  ) {
    // for a comment based assert specification the location is essentially
    // the start of the comment to the end of the assert.
    exprComment.getLocation().hasLocationInfo(filepath, startline, startcolumn, _, _) and
    this.(LibcAssert).getLocation().hasLocationInfo(_, _, _, endline, endcolumn)
  }

  string getContents() { result = contents }

  override string getKeyword() { result = keyword }
}
