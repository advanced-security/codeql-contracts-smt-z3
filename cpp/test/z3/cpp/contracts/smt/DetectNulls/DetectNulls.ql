/**
 * @id cpp/test/SMT/DetectNulls
 * @name DetectsNulls
 * @description Detects nullity 
 * @kind problem
 * @precision high
 * @problem.severity warning
 */

import cpp
import z3.cpp.contracts.SpecificationCase
import z3.cpp.contracts.SMT
import z3.cpp.contracts.style.AnnotatedCAssertStyleSpecifiedElement

// Tests to see if an expression should be considered to be null. 
from CodeQLSpecificationClauseRequires requires, Expr e, VariableAccess va
  where 
      e = requires.getClauseExpr() and 
      isSupported(e) and
      va.getParent*() = e 
      and possiblyNull(va)
select requires, "Within contract $@ is possibly null", va, "this variable"  
  