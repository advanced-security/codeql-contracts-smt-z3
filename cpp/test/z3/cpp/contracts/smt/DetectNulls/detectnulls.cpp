#include "assert.h"
#include <cstddef>

class A {
public:
    int *a;
    int d;
    int operator==(A x)
    {
        return 0;
    }
};
 
void f1(int *a, int b){
    /*@ requires @*/
    assert(a);

    /*@ requires @*/
    assert(a && b > 0);

}

void f2(int *a){
    /*@ requires @*/
    assert(a != NULL);
}

void f4(A *a){
    /*@ requires @*/
    assert(a);
}

void f5(A *a){
    /*@ requires @*/
    assert(a != NULL);
}

// tricky case because they may pass the object and not 
// the field
void f6(A &b){
    /*@ requires @*/
    assert(b.a);

    /*@ requires @*/
    assert(b.a != NULL);
}


void t1(){
    int *a;
    int b;

    f1(a, b);

    if(a != NULL){
        f1(a,b);
    }else{
        f1(a,b);
    }

    if(a == NULL){
        f1(a,b);
    }
}

void t2(){
    int b;
    int *a = &b;

    f1(a, b);

    if(a != NULL){
        f1(a,b);
    }else{
        f1(a,b);
    }

    if(a == NULL){
        f1(a,b);
    }
}

void t3(){
    int b;
    int *a = &b;

    f1(a, b);

    if(a != NULL){
        f1(a,b);
    }else{
        f1(a,b);
    }

    f1(a,b);
}


void t4(){
    int b;
    int *a;

    f1(a, b);

    if(a != NULL){
        f1(a,b);
    }else{
        f1(a,b);
    }

    f1(a,b);
}

