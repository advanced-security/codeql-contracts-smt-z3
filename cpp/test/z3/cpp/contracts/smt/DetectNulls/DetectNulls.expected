| detectnulls.cpp:15:5:16:13 | requires a | Within contract $@ is possibly null | detectnulls.cpp:16:12:16:12 | a | this variable |
| detectnulls.cpp:18:5:19:22 | requires ... && ... | Within contract $@ is possibly null | detectnulls.cpp:19:12:19:12 | a | this variable |
| detectnulls.cpp:18:5:19:22 | requires ... && ... | Within contract $@ is possibly null | detectnulls.cpp:19:17:19:17 | b | this variable |
| detectnulls.cpp:24:5:25:21 | requires ... != ... | Within contract $@ is possibly null | detectnulls.cpp:25:12:25:12 | a | this variable |
| detectnulls.cpp:29:5:30:13 | requires a | Within contract $@ is possibly null | detectnulls.cpp:30:12:30:12 | a | this variable |
| detectnulls.cpp:34:5:35:21 | requires ... != ... | Within contract $@ is possibly null | detectnulls.cpp:35:12:35:12 | a | this variable |
| detectnulls.cpp:41:5:42:15 | requires a | Within contract $@ is possibly null | detectnulls.cpp:42:12:42:12 | b | this variable |
| detectnulls.cpp:41:5:42:15 | requires a | Within contract $@ is possibly null | detectnulls.cpp:42:14:42:14 | a | this variable |
| detectnulls.cpp:44:5:45:23 | requires ... != ... | Within contract $@ is possibly null | detectnulls.cpp:45:12:45:12 | b | this variable |
| detectnulls.cpp:44:5:45:23 | requires ... != ... | Within contract $@ is possibly null | detectnulls.cpp:45:14:45:14 | a | this variable |
